packages/aws-cdk-lib/aws-ecs/lib/credential-spec.ts

```diff
@@ -0,0 +1,139 @@
+import { IBucket } from '../../aws-s3';
+import { IParameter } from '../../aws-ssm';
+
+/**
+ * Base construct for a credential specification (CredSpec).
+ */
+export class CredentialSpec {
+  /**
+   * Get the ARN for an S3 object.
+   */
+  protected static arnForS3Object(bucket: IBucket, key: string) {
+    if (!key) {
+      throw new Error('key is undefined');
+    }
+
+    return bucket.arnForObjects(key);
+  }
+
+  /**
+   * Get the ARN for a SSM parameter.
+   */
+  protected static arnForSsmParameter(parameter: IParameter) {
+    return parameter.parameterArn;
+  }
+
+  /**
+   * Prefix string based on the type of CredSpec.
+   */
+  public prefixId: string;
+
+  /**
+   * Location or ARN from where to retrieve the CredSpec file.
+   */
+  public fileLocation: string;
+
+  /**
+   * @param fileLocation Location or ARN from where to retrieve the CredSpec file
+   */
+  constructor(prefixId: string, fileLocation: string) {
+    this.prefixId = prefixId;
+    this.fileLocation = fileLocation;
+  }
+
+  /**
+   * Called when the container is initialized to allow this object to bind
+   * to the stack.
+   */
+  public bind(): CredentialSpecConfig {
+    return {
+      typePrefix: this.prefixId,
+      location: this.fileLocation,
+    };
+  }
+}
+
+/**
+ * Credential specification (CredSpec) file.
+ */
+export class DomainJoinedCredentialSpec extends CredentialSpec {
+  /**
+   * Prefix Id for this type of CredSpec.
+   */
+  public static readonly PREFIX_ID = 'credentialspec';
+
+  /**
+   * Loads the CredSpec from a S3 bucket object.
+   *
+   * @param bucket The S3 bucket
+   * @param key The object key
+   * @returns CredSpec with it's locations set to the S3 object's ARN.
+   */
+  public static fromS3Bucket(bucket: IBucket, key: string) {
+    return new DomainJoinedCredentialSpec(CredentialSpec.arnForS3Object(bucket, key));
+  }
+
+  /**
+   * Loads the CredSpec from a SSM parameter.
+   *
+   * @param parameter The SSM parameter
+   * @returns CredSpec with it's locations set to the SSM parameter's ARN.
+   */
+  public static fromSsmParameter(parameter: IParameter) {
+    return new DomainJoinedCredentialSpec(CredentialSpec.arnForSsmParameter(parameter));
+  }
+
+  constructor(fileLocation: string) {
+    super(DomainJoinedCredentialSpec.PREFIX_ID, fileLocation);
+  }
+}
+
+/**
+ * Credential specification for domainless gMSA.
+ */
+export class DomainlessCredentialSpec extends CredentialSpec {
+  /**
+   * Prefix Id for this type of CredSpec.
+   */
+  public static readonly PREFIX_ID = 'credentialspecdomainless';
```

Comment:
same here. I think we can just put the raw string in the constructor

Reply:
done, moved to constructor